airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -123,6 +123,51 @@ def read_full_refresh(
                 cursor_field=cursor_field,
             )
 
+    def read_incremental(
+        self,
+        cursor_field: Optional[List[str]],
+        logger: logging.Logger,
+        slice_logger: SliceLogger,
+        stream_state: MutableMapping[str, Any],
+        state_manager,  # ignoring typing for ConnectorStateManager because of circular dependencies
+        per_stream_state_enabled: bool,
+    ) -> Iterable[StreamData]:
+        slices = self.stream_slices(
+            cursor_field=cursor_field,
+            sync_mode=SyncMode.incremental,
+            stream_state=stream_state,
+        )
+        logger.debug(f"Processing stream slices for {self.name} (sync_mode: incremental)", extra={"stream_slices": slices})
+
+        has_slices = False
+        record_counter = 0
+        for _slice in slices:
+            has_slices = True
+            if slice_logger.should_log_slice_message(logger):
+                yield slice_logger.create_slice_log_message(_slice)
+            records = self.read_records(
+                sync_mode=SyncMode.incremental,
+                stream_slice=_slice,
+                stream_state=stream_state,
+                cursor_field=cursor_field or None,
+            )
+            for record_data_or_message in records:
+                yield record_data_or_message
+                if isinstance(record_data_or_message, Mapping) or record_data_or_message.type == MessageType.RECORD:
+                    record_data = record_data_or_message if isinstance(record_data_or_message, Mapping) else record_data_or_message.record
+                    stream_state = self.get_updated_state(stream_state, record_data)
+                    checkpoint_interval = self.state_checkpoint_interval
+                    record_counter += 1
+                    if checkpoint_interval and record_counter % checkpoint_interval == 0:
```

Comment:
Previous implementation was using the number of records in a slice. This has been changed here to consider the records across all the slices